tensorpipe/test/channel/channel_test.cc

```diff
@@ -15,513 +15,568 @@
+// correlated one: that the recv callback isn't called inline from within the
+// recv method. We do so by having that behavior cause a deadlock.
+class CallbacksAreDeferredTest : public ChannelTest<tensorpipe::CpuTensor> {
+  static constexpr auto dataSize = 256;
+
+  void server(std::shared_ptr<transport::Connection> conn) override {
+    std::shared_ptr<CpuContext> ctx = this->helper_->makeContext("server");
+    auto channel = ctx->createChannel(std::move(conn), Endpoint::kListen);
+
+    // Initialize with sequential values.
+    std::vector<uint8_t> data(dataSize);
+    std::iota(data.begin(), data.end(), 0);
+
+    // Perform send and wait for completion.
+    std::promise<std::tuple<Error, TDescriptor>> descriptorPromise;
+    std::promise<Error> sendPromise;
+    std::mutex mutex;
+    std::unique_lock<std::mutex> callerLock(mutex);
+    channel->send(
+        CpuTensor{
+            .ptr = data.data(),
+            .length = data.size(),
+        },
+        [&descriptorPromise](const Error& error, TDescriptor descriptor) {
+          descriptorPromise.set_value(
+              std::make_tuple(error, std::move(descriptor)));
+        },
+        [&sendPromise, &mutex](const Error& error) {
+          std::unique_lock<std::mutex> calleeLock(mutex);
+          sendPromise.set_value(error);
+        });
+    callerLock.unlock();
+    Error descriptorError;
+    TDescriptor descriptor;
+    std::tie(descriptorError, descriptor) =
+        descriptorPromise.get_future().get();
+    EXPECT_FALSE(descriptorError) << descriptorError.what();
+    this->peers_->send(PeerGroup::kClient, descriptor);
+    Error sendError = sendPromise.get_future().get();
+    EXPECT_FALSE(sendError) << sendError.what();
+
+    this->peers_->done(PeerGroup::kServer);
+    this->peers_->join(PeerGroup::kServer);
+
+    ctx->join();
+  }
+
+  void client(std::shared_ptr<transport::Connection> conn) override {
+    std::shared_ptr<CpuContext> ctx = this->helper_->makeContext("client");
+    auto channel = ctx->createChannel(std::move(conn), Endpoint::kConnect);
+
+    // Initialize with zeroes.
+    std::vector<uint8_t> data(dataSize);
+    std::fill(data.begin(), data.end(), 0);
+
+    // Perform recv and wait for completion.
+    std::promise<Error> recvPromise;
+    std::mutex mutex;
+    std::unique_lock<std::mutex> callerLock(mutex);
+    auto descriptor = this->peers_->recv(PeerGroup::kClient);
+    channel->recv(
+        descriptor,
+        CpuTensor{
+            .ptr = data.data(),
+            .length = data.size(),
+        },
+        [&recvPromise, &mutex](const Error& error) {
+          std::unique_lock<std::mutex> calleeLock(mutex);
+          recvPromise.set_value(error);
+        });
+    callerLock.unlock();
+    Error recvError = recvPromise.get_future().get();
+    EXPECT_FALSE(recvError) << recvError.what();
+
+    // Validate contents of vector.
+    for (auto i = 0; i < dataSize; i++) {
```

Comment:
Also a reference to a previous comment: can we stick to dataSize or data.size() consistently?